text/0011-peer-chaincode-optimization.md

```diff
@@ -0,0 +1,265 @@
+
+[1]: https://github.com/hyperledger/fabric-rfcs/pull/58#issuecomment-2458605718
+[2]: https://github.com/hyperledger/fabric-rfcs/pull/58#issuecomment-2460109295
+[3]: https://github.com/hyperledger/fabric-rfcs/pull/58#issuecomment-2473626315
+
+
+### Proposed Solutions
+
+1. Maintaining Read-Write Order
+
+	•	Solution: Ensure that the batching mechanism processes operations in the same order as they are invoked in the chaincode. This approach can be achieved by batching only consecutive write operations and sending read operations interactively.
+	•	Implementation: The chaincode shim can collect PutState or DelState operations into a batch, but any GetState operation will immediately flush the current batch and process the read interactively. This guarantees consistency with the current behavior.
+
+2. Handling Dependencies and Validation
+
+	•	Solution: Retain validation logic (e.g., metadata reads and collection name checks) on the peer side to ensure correctness. Avoid shifting these responsibilities to the shim, as it increases complexity and risk.
+	•	Implementation: The peer should process batched operations in a way that preserves existing checks and validations, ensuring no functional changes to current read-write dependencies.
+
+3. Explicit Developer Control
+
+	•	Solution: Introduce explicit APIs for developers to enable and manage batching in their chaincode. This gives developers control over when and how batching is applied, allowing for easier debugging and incremental adoption.
+	•	Implementation: APIs like StartBatch and FinishBatch can provide clear demarcation points for batching operations, ensuring developers understand and control the batching process.
+
+4. Optimizing Interleaved Operations
+
+	•	Solution: Allow partial batching where consecutive writes are batched, and reads or other operations trigger immediate processing of pending batches. This minimizes communication overhead without significantly altering the behavior for interleaved read-write scenarios.
+	•	Implementation: The shim can dynamically manage batches and flush them when necessary, ensuring optimal performance while maintaining behavior consistency.
+
+## Final Solution
+
+### Extend Chaincode Stub API with Explicit Batching APIs
+
+Introduce new APIs in the chaincode stub to explicitly control batching operations:
+
+```go
+// StartWriteBatch initializes a new batch for write operations.
+func (stub *ChaincodeStub) StartWriteBatch() {
+    // Implementation to begin batching write operations.
+}
+
+// FinishWriteBatch flushes the current batch and sends all collected write operations to the peer.
+// Returns a serialized set of results for each operation with the corresponding status.
+func (stub *ChaincodeStub) FinishWriteBatch() ([]BatchResult, error) {
```

Comment:
Sorry, but I'm confused by []BatchResult.
What will be written there? 
Usually a change operation returns an error or nil. 
Batch will be applied on peer sequentially. And if at least one operation will end with an error, then immediately the processing of the whole batch will end with an error. 
Therefore, either []BatchResult will consist of all nil, or some number of nil and one error.

Wouldn't it be better to replace []BatchResult with error?

Reply:
That was only a suggestion, I think you have a point and having only an error should be indeed enough of indication, given that implementation wise we won't continue to try and execute batch once we stumbled into the error.

Reply:
I agree a simple error will suffice, returned when the first error is hit.